### PR TITLE
8256415: Shenandoah: Should evacuate/update codecache concurrently when class unloading is off

### PR DESCRIPTION
**Pull request by zhengyu123**
During isolating GC work (JDK-8255765), I found that Shenandoah still evacuate/update codecache roots at final mark safepoint when class unloading is off.

Actually, it can be done concurrently by utilizing nmethod_entry_barrier.

Shenandoah should always enable nmethod_entry_barrier for evacuation, regardless class unloading status, and perform code cache evacuation/updating during strong root processing.

Key points:
1) Always enable nmethod entry barrier
2) Always attach ShenandoahNMethod to nmethod, regardless weather class unloading is enabled
3) Only evacuate/update thread and serial weak roots at final mark pause.
4) Code roots are evacuated/updated concurrently during weak roots (when class unloading is on) or strong roots (when class unloading is off)

- [x] hotspot_gc_shenandoah
- [x] tier1 with ShenandoahGC and +/-ClassUnloading

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8256415](https://bugs.openjdk.java.net/browse/JDK-8256415): Shenandoah: Should evacuate/update codecache concurrently when class unloading is off

### Reviewers
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to 5e96a2c55a57d9532384f2cb576641ad1e4b56a1

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1271/head:pull/1271`
`$ git checkout pull/1271`

### CODE REVIEW

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp [hunk 1]
```diff
@@ -1744,9 +1742,9 @@ void ShenandoahHeap::op_final_mark() {
       set_has_forwarded_objects(true);
 
       if (!is_degenerated_gc_in_progress()) {
-        if (ShenandoahConcurrentRoots::should_do_concurrent_class_unloading()) {
-          ShenandoahCodeRoots::arm_nmethods();
-        }
+        // We need to arm nmethods for concurrent codecache process, regardless concurrent
+        // class unloading
```
Comment (rkennke) [inline src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:1746]: Comment should probably not refer to previous implementations. Future readers will not see the connection there.

**Review by rkennke — APPROVED** [review]
Looks good, except change one comment a little.

File: src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp (hunk 1 above)
Comment (zhengyu123) [reply src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp]: Thanks for reviewing.

Comment updated to "Arm nmethods for concurrent codecache processing."

**Pull request closed without merging**